rfd/00XX-proxy-tunnel-service.md

```diff
@@ -0,0 +1,122 @@
+---
+authors: David Boslee (david@goteleport.com), Naji Obeid (naji@goteleport.com)
+state: draft
+---
+
+# RFD XX - Proxy Tunnel Service
+
+## What
+This document describes an API that enables a proxy to dial the nodes connected to a neighboring proxy. This will allow node agents to connect to one proxy and be reachable through any other proxy in the cluster. This is one component of a larger design to make teleport more cloud friendly. See [RFD 48](https://github.com/gravitational/teleport/blob/master/rfd/0048-war-dialler-node-tracker.md) for more details.
+
+## Why
+The main goal is to remove the need for a node agent to create a reverse tunnel to every proxy. The problems caused by this behavior are outlined [here](https://github.com/gravitational/teleport/blob/master/rfd/0048-war-dialler-node-tracker.md#why).
+
+## Details
+
+### Proxy API
+
+The following gRPC service will be added to proxy servers:
+
+```protobuf
+syntax = "proto3";
+
+package api;
+
+service TunnelService { rpc DialNode(stream Frame) returns (stream Frame); }
+
+// Frame wraps different message types to be sent over a stream.
+message Frame {
+   oneof Message {
+       DialRequest DialRequest = 1;
+       Data Data = 2;
+   }
+}
+
+// DialRequest contains details for connecting to a node.
+message DialRequest {
+   // ServerID is the {UUID}.{ClusterName} of a node.
+   string ServerID = 1;
+   // ConnType is the type of connection requested.
+   // Examples: “node” or “app”
+   string ConnType = 2;
+   // Source is the source address of the connection.
+   Addr Source = 3;
+   // Destination is the destination address of the connection.
+   Addr Destination = 4;
+}
+
+message Addr {
+   // Network is the name of the network transport.
```

Comment:
This is just a net.Addr interface passed over grpc. I actually am not certain if teleport puts custom values here but this would something like "tcp" or "udp". It doesn't seem like its actually needed anywhere but I added it to construct the net.Addr consistently on the rt-proxy side of things.